src/Mvc/Mvc.Core/src/Routing/ControllerActionEndpointDataSource.cs

```diff
@@ -108,6 +108,28 @@ protected override List<Endpoint> CreateEndpoints(IReadOnlyList<ActionDescriptor
 
             return endpoints;
         }
+
+        internal void AddDynamicControllerEndpoint(IEndpointRouteBuilder endpoints, string pattern, Type transformerType, object state)
+        {
+            CreateInertEndpoints = true;
+            lock (Lock)
+            {
+                var order = _orderSequence.GetNext();
+
+                endpoints.Map(
+                    pattern,
+                    context =>
+                    {
+                        throw new InvalidOperationException("This endpoint is not expected to be executed directly.");
+                    })
+                    .Add(b =>
+                    {
+                        ((RouteEndpointBuilder)b).Order = order;
```

Comment:
This looks like a breaking change since users can no longer provide a custom builder. Can we continue supporting the interface?

Reply:
They can't in any case, we already depend on this behavior in all the MapFallback**** overloads, so this is not an issue.